tokio/src/io/util/read_to_end.rs

```diff
@@ -73,70 +72,53 @@ unsafe fn poll_read_to_end<R: AsyncRead + ?Sized>(
-        // never expose uninitialized memory in the vector.
-        //
-        // Note that the assert! below ensures that we don't set the length to
-        // something larger than the capacity, which malicious implementors might
-        // try to have us do.
-        //
-        // 2. The AsyncRead has overriden `prepare_uninitialized_buffer`.
-        //
-        // In this case, the safety of the `set_len` call below relies on this
-        // guarantee from the documentation on `prepare_uninitialized_buffer`:
-        //
-        // > This function isn't actually unsafe to call but unsafe to implement.
-        // > The implementer must ensure that either the whole buf has been zeroed
-        // > or poll_read() overwrites the buffer without reading it and returns
-        // > correct value.
-        //
-        // Note that `prepare_uninitialized_buffer` is unsafe to implement, so this
-        // is a guarantee we can rely on in unsafe code.
-        //
-        // The assert!() is technically only necessary in the first case.
-        let new_len = buf.len() + num;
-        assert!(new_len <= buf.capacity());
-
-        buf.set_len(new_len);
-    }
-    Poll::Ready(res)
-}
-
-/// This function prepares the unused capacity for use with the provided AsyncRead.
-pub(super) fn prepare_buffer<R: AsyncRead + ?Sized>(buf: &mut Vec<u8>, read: &R) {
-    let buffer = get_unused_capacity(buf);
-
-    // safety: This function is only unsafe to implement.
-    unsafe {
-        read.prepare_uninitialized_buffer(buffer);
-    }
+    reserve(buf, 32);
+
+    let mut unused_capacity = ReadBuf::uninit(get_unused_capacity(buf));
+
+    ready!(read.poll_read(cx, &mut unused_capacity))?;
+
+    // safety: There are two situations:
+    //
+    // 1. The AsyncRead has not overriden `prepare_uninitialized_buffer`.
+    //
+    // In this situation, the default implementation of that method will have
+    // zeroed the unused capacity. This means that setting the length will
+    // never expose uninitialized memory in the vector.
+    //
+    // Note that the assert! below ensures that we don't set the length to
+    // something larger than the capacity, which malicious implementors might
+    // try to have us do.
+    //
+    // 2. The AsyncRead has overriden `prepare_uninitialized_buffer`.
+    //
+    // In this case, the safety of the `set_len` call below relies on this
+    // guarantee from the documentation on `prepare_uninitialized_buffer`:
+    //
+    // > This function isn't actually unsafe to call but unsafe to implement.
+    // > The implementer must ensure that either the whole buf has been zeroed
+    // > or poll_read() overwrites the buffer without reading it and returns
+    // > correct value.
+    //
+    // Note that `prepare_uninitialized_buffer` is unsafe to implement, so this
+    // is a guarantee we can rely on in unsafe code.
+    //
+    // The assert!() is technically only necessary in the first case.
```

Comment:
also, AFAICT, this function was unsafe to call because it required `prepare_uninitialized_buffer` to be called. now, it doesn't, since that no longer exists. does it make sense to make it safe to call this function, and scope the `unsafe` block only to the `Vec::set_len` call? the rest of this code appears to be safe.